packages/destination-actions/src/destinations/avo/sendSchemaToInspector/avo.ts

```
@@ -29,27 +57,189 @@ function generateBaseBody(event: Payload, appVersionPropertyName: string | undef
+function handleEvent(
+  baseBody: BaseBody,
+  event: Payload,
+  eventProperties: EventProperty[],
+  metadata?: EventSpecMetadata
+): EventSchemaBody {
   // Initially declare eventBody with the type EventSchemaBody
   // and explicitly set all properties to satisfy the type requirements.
   const eventBody: EventSchemaBody = {
     ...baseBody, // Spread operator to copy properties from baseBody
     type: 'event', // Explicitly set type as 'event'
     eventName: event.event, // Set from the event parameter
-    eventProperties: AvoSchemaParser.extractSchema(event.properties),
+    eventProperties: eventProperties, // Use the provided event properties (may include validation results)
     eventId: null, // Set default or actual value
-    eventHash: null // Set default or actual value
+    eventHash: null, // Set default or actual value
+    metadata: metadata // Include validation metadata if present
   }
 
   return eventBody
 }
 
-export function extractSchemaFromEvent(event: Payload, appVersionPropertyName: string | undefined) {
-  const baseBody: BaseBody = generateBaseBody(event, appVersionPropertyName)
+/**
+ * Fetches the event spec if spec fetching is enabled (schemaId/sourceId provided).
+ * Returns a Promise that resolves to an object with the event spec (or null) and debug info.
+ *
+ * Note: Spec fetching happens regardless of encryption key presence.
+ * The encryption key only controls whether property values are sent (Phase 2).
+ * If spec fetch fails (invalid status code), validation is skipped for that event.
+ */
+async function fetchEventSpec(
+  eventName: string,
+  apiKey: string,
+  streamId: string | undefined,
+  env: string,
+  request: RequestClient
+): Promise<EventSpecResponse | null> {
+  if (!apiKey) {
+    console.warn(`[Avo Inspector] apiKey is missing, cannot fetch event spec`)
+    return null
+  }
+
+  if (!streamId) {
+    console.warn(`[Avo Inspector] streamId is missing, cannot fetch event spec`)
```

We don't have the ability to log warnings. Please throw an error or decide to proceed.